### PR TITLE
[TR] Clean behat grid context

### PR DESCRIPTION
**Pull request by phaseinducer**
The goal of this PR is to improve the way the grid is tested in our behats by :
- reworking the methods that fetch information inside the grid (speed improvement)
- adding more spins (stability improvement)
- removing lots of `catch (\Exception $e)` (debuggability improvement)
- fixing a false positive case when sorting by a boolean column
- cleaning phpdoc

For example, the sorting steps ("I should be able to sort the rows by...") are around 4x faster with this implementation. Although the gain is significant on a local machine, it does not really speed up the builds on the CI (don't know why : / ).

CI is blue.

### CODE REVIEW

**Comment by jjanvier** [comment]
GTM :+1:

File: features/Context/DataGridContext.php [hunk 1]
```diff
@@ -497,7 +497,6 @@ public function pageSizeShouldBe($size)
      */
     public function iSortByValue($columnName, $order = 'ascending')
     {
-        $this->wait(10000, sprintf('$("a:contains(\'%s\')").length > 0', ucfirst($columnName)));
         $this->datagrid->sortBy($columnName, $order);
         $this->wait();
```
Comment (ClementGautier) [inline features/Context/DataGridContext.php]: could we remove this one too ? we could change it for a spin for example

File: features/Context/Page/Base/Grid.php [hunk 2]
```diff
@@ -742,28 +716,25 @@ protected function deactivateFilter($filterName)
...
+            function () use ($filterName) {
                 $filterElement = $this
                     ->getElement('Manage filters')
                     ->find('css', sprintf('label:contains("%s")', $filterName));
-
-                if ($filterElement) {
+                if (null !== $filterElement) {
                     $filterElement->click();
 
                     return true;
```
Comment (ClementGautier) [inline features/Context/Page/Base/Grid.php]: easier I guess:
```php
$node = $this->spin(function () use ($filterName) {
    return $this->getElement('Manage filters')->find('css', sprintf('label:contains("%s")', $filterName));
}, 20, sprintf('Impossible to find filter "%s"', $filterName));

$node->->click();
```

**Comment by grena** [comment]
Nice! GTM

File: features/Context/Page/Base/Grid.php [hunk 3]
```diff
@@ -771,18 +742,22 @@ protected function clickOnFilterToManage($filterName)
...
-                $filterList->click();
+                if (null !== $filterList) {
+                    $filterList->click();
 
-                return true;
-            });
-        } catch (\Exception $e) {
-            throw new \InvalidArgumentException('Impossible to find filter list');
-        }
+                    return true;
```
Comment (ClementGautier) [inline features/Context/Page/Base/Grid.php]: same as clickOnFilterToManage

File: features/Context/Page/Base/Grid.php [hunk 4]
```diff
@@ -1089,14 +1041,16 @@ public function sequentialEdit()
...
-     * @throws \InvalidArgumentException
      */
     public function selectAll()
     {
-        if (!$allBtn = $this->getDropdownSelector()->find('css', 'button:contains("All")')) {
-            throw new \InvalidArgumentException('"All" button on dropdown row selector not found');
-        }
+        $allBtn = $this->spin(
+            function () {
+                return $this->getDropdownSelector()->find('css', 'button:contains("All")');
```
Comment (ClementGautier) [inline features/Context/Page/Base/Grid.php]: you can extract $this->getDropdownSelector() from the spin to be faster:
```
$selector  = $this->getDropdownSelector();
$button = $this->spin(
    function () use ($selector) {return $selector->find('css', 'button:contains("All")');}),
    20,
    '"All" button on dropdown row selector not found'
);
```

File: features/Context/Page/Base/Grid.php [hunk 5]
```diff
@@ -1120,37 +1074,42 @@ public function selectNone()
...
      */
     protected function clickOnDropdownSelector($item)
     {
-        if (!$dropdown = $this->getDropdownSelector()->find('css', 'button.dropdown-toggle')) {
-            throw new \InvalidArgumentException('Dropdown row selector not found');
-        }
-
+        $dropdown = $this->spin(
+            function () {
+                return $this->getDropdownSelector()->find('css', 'button.dropdown-toggle');
```
Comment (ClementGautier) [inline features/Context/Page/Base/Grid.php]: same

File: features/Context/Page/Base/Grid.php (hunk 2 above)
Comment (phaseinducer) [inline features/Context/Page/Base/Grid.php]: :+1:

**Comment by ClementGautier** [comment]
GTM after small changes

File: features/Context/Page/Base/Grid.php [hunk 6]
```diff
@@ -412,34 +411,37 @@ public function getCellImage($column, $row)
...
+     * @param string $columnName
+     * @param bool   $withHidden
      * @param bool   $withActions
      *
+     * @throws \InvalidArgumentException
+     *
      * @return int
      */
-    public function getColumnPosition($column, $withActions = false)
+    public function getColumnPosition($columnName, $withHidden, $withActions)
```
Comment (jmleroux) [inline features/Context/Page/Base/Grid.php]: you should keep `$withActions` in second position.
And add a default value for `$withHidden`

File: features/Context/Page/Base/Grid.php [hunk 7]
```diff
@@ -486,48 +487,23 @@ public function countColumns()
...
-        }
-
-        return $sorter;
+        $header = $this->getColumnHeader($columnName);
+
+        return $this->spin(
+            function () use ($header) {
+                return $header->find('css', 'a');
+            },
+            20,
```
Comment (jmleroux) [inline features/Context/Page/Base/Grid.php]: this parameter will disappear soon...

File: features/Context/Page/Base/Grid.php (hunk 2 above)
Comment (jmleroux) [inline features/Context/Page/Base/Grid.php]: spin should almost never click

File: features/Context/Page/Base/Grid.php (hunk 6 above)
Comment (phaseinducer) [inline features/Context/Page/Base/Grid.php]: No problem, I checked all usages.

File: features/Context/Page/Base/Grid.php (hunk 7 above)
Comment (phaseinducer) [inline features/Context/Page/Base/Grid.php]: To remove on master

**Comment by nidup** [comment]
:santa:

**Pull request merged**